File: Section06OptionalChallenges/DivideAndConquer_findRotatedIndex.js
function findRotatedIndex(arr, num) {
  let index = findNumInRotatedArray(arr, num);
  return index < 0 ? -1 : index;
}

function findNumInRotatedArray(arr, num, lowIndex = 0, highIndex = arr.length - 1) {
  if (highIndex >= lowIndex) {
    let midIndex = Math.floor((highIndex + lowIndex) / 2);
    if (arr[midIndex] === num) {
      return midIndex;
    }
    let leftIndex = findNumInRotatedArray(arr, num, lowIndex, midIndex - 1);
    if (leftIndex >= 0) {
      return leftIndex;
    }
    let rightIndex = findNumInRotatedArray(arr, num, midIndex + 1, highIndex);
    if (rightIndex >= 0) {
      return rightIndex;
    }
  }
  return -1;
}

console.log(findRotatedIndex([3,4,1,2],4));
console.log(findRotatedIndex([6, 7, 8, 9, 1, 2, 3, 4], 8));
console.log(findRotatedIndex([6, 7, 8, 9, 1, 2, 3, 4], 3));
console.log(findRotatedIndex([37,44,66,102,10,22],14));
console.log(findRotatedIndex([6, 7, 8, 9, 1, 2, 3, 4], 12));
console.log(findRotatedIndex([11,12,13,14,15,16,3,5,7,9], 16));

File: Section06OptionalChallenges/Section06OptionalChallenges_DOTNET/Section06OptionalChallenges/DivideAndConquer_FindRotatedIndex.cs
using System;

namespace Section06OptionalChallenges.DivideAndConquer;

public partial class DivideAndConquer
{
  public static int FindRotatedIndex(int[] arr, int num)
  {
    var index = FindNumInRotatedArray(arr, num, 0, arr.Length - 1);
    return index < 0 ? -1 : index;
  }

  private static int FindNumInRotatedArray(int[] arr, int num, int lowIndex, int highIndex)
  {
    if (highIndex >= lowIndex)
    {
      var midIndex = (int)Math.Floor((double)(highIndex + lowIndex) / 2);
      if (arr[midIndex] == num)
      {
        return midIndex;
      }
      var leftIndex = FindNumInRotatedArray(arr, num, lowIndex, midIndex - 1);
      if (leftIndex >= 0)
      {
        return leftIndex;
      }
      var rightIndex = FindNumInRotatedArray(arr, num, midIndex +1, highIndex);
      if (rightIndex >= 0)
      {
        return rightIndex;
      }
    }
    return -1;
  }
}